### PR TITLE
Automaton shift declaration update

### PR DESCRIPTION
**Pull request by chernyiv**
**I. New parser rules were added:**
    1.     Shift declaration with 'by' token and only one function name without any brackets (if only one function provides shift)
    2.     Shift declaration with 'by' token, square brackets and trailing comma (if many functions are used)
    3.     Two similar rules for cases when more than one state can be changed
    4.     Initial ways to declare shifts were not removed, but there were some changes in nodes in tests _(see next paragraphs)_
   
**II. Shift data class in nodes package was extended:**
    The `dumpString()` method now adds functions arg type names to the tree.
    
**III. Some testing changes:**
    1.    New shift declarations are tested with use of `shiftSquareBrackets.lsl` file     
    2.    `simple.lsl` file was changed because argument types of functions can be printed now (they were not expected before)

### CODE REVIEW

**Comment by vldF** [comment]
Please, change the target branch to the `develop`

File: src/main/antlr/LibSLParser.g4 [hunk 1]
```diff
@@ -129,12 +129,16 @@ automatonStateDecl
...
+   |   SHIFT from=Identifier MINUS_ARROW to=Identifier BY L_SQUARE_BRACKET functionsList? R_SQUARE_BRACKET SEMICOLON
+   |   SHIFT from=Identifier MINUS_ARROW to=Identifier L_BRACKET functionsList? R_BRACKET SEMICOLON
+   |   SHIFT from=L_BRACKET identifierList R_BRACKET MINUS_ARROW to=Identifier  BY functionsListPart SEMICOLON
    |   SHIFT from=L_BRACKET identifierList R_BRACKET MINUS_ARROW to=Identifier L_BRACKET functionsList? R_BRACKET SEMICOLON
+   |   SHIFT from=L_BRACKET identifierList R_BRACKET MINUS_ARROW to=Identifier BY L_SQUARE_BRACKET functionsList? R_SQUARE_BRACKET SEMICOLON
    ;
 
 functionsList
-   :   functionsListPart (COMMA functionsListPart)*
+   :   functionsListPart (COMMA functionsListPart)* (COMMA)?
```
Comment (vldF) [inline src/main/antlr/LibSLParser.g4:139]: Yeah, this way to support trailing comma looks easier

**Comment by vldF** [comment]
> Initial ways to declare shifts were not removed

May be you should remove them? To make lsl-code cleaner

File: src/test/testdata/lsl/shiftSquareBrackets.lsl [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    shift S1 -> S2 by foo;
+    shift S1 -> S2(fix);
+    shift S1 -> S2 by [fix, foo,];
+    shift S1 -> S2 by [foo];
+    shift S1 -> S2 by [foo,];
+    shift S1 -> S3 by [sum(int, int), min(int), ];
+    shift S1 -> S3 by sum(int, int);
+
+    shift (S1, S2) -> S3 by foo;
+    shift (S1, S2) -> S3 by [sum(int, int), min(int),];
```
Comment (vldF) [inline src/test/testdata/lsl/shiftSquareBrackets.lsl]: Could you add at least one test with the new syntax (with brackets) with 2+ args and without trailing comma?

Or you could remove the comma from this line

Comment (chernyiv) [reply src/test/testdata/lsl/shiftSquareBrackets.lsl]: Done!

File: src/test/testdata/expected/lsl/shiftSquareBrackets.lsl [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+}
+automaton a : int {
+    state S1;
+    state S2;
+    state S3;
+    shift S1 -> S2(foo);
+    shift S1 -> S2(fix, foo);
+    shift S1 -> S2(foo);
+    shift S1 -> S2(foo);
+    shift S1 -> S3(sum(int, int), min(int));
```
Comment (vldF) [inline src/test/testdata/expected/lsl/shiftSquareBrackets.lsl]: It looks like you didn't update `dumpToString()` function to drop out old syntax

File: src/test/testdata/lsl/shiftSquareBrackets.lsl [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+    shift (S1, S2) -> S3 by foo;
+    shift (S1, S2) -> S3 by [sum(int, int), min(int)];
+
+
+    fun foo();
+    fun sum(i: int, j: int);
+    fun min(i: int);
+    fun fix();
+}
```
Comment (vldF) [inline src/test/testdata/lsl/shiftSquareBrackets.lsl]: Please, add an empty line at the end of file. And enable automatically trailing line breaks in Settings (ctrl+alt+S)

image

File: src/test/testdata/expected/lsl/shiftSquareBrackets.lsl (hunk 3 above)
Comment (chernyiv) [reply src/test/testdata/expected/lsl/shiftSquareBrackets.lsl]: Updated the function, but there is a question. Should we check  for trailing comma in functions list (print it out too) and absence of square brackets (if there is only one function)? 
Right now `dumpToString()` always prints square brackets and doesn't check for trailing comma.

File: src/test/testdata/lsl/shiftSquareBrackets.lsl (hunk 4 above)
Comment (chernyiv) [reply src/test/testdata/lsl/shiftSquareBrackets.lsl]: Done

File: src/test/testdata/expected/lsl/shiftSquareBrackets.lsl (hunk 3 above)
Comment (vldF) [reply src/test/testdata/expected/lsl/shiftSquareBrackets.lsl]: This always depends on you. I think the better way is to absence square brackets and to skip trailing comma

**Review by vldF — APPROVED** [review]

**Pull request merged**